nefertari/es_query.py

```diff
@@ -0,0 +1,329 @@
+
+    if _is_nested(item):
+        _attach_nested(item, aggregation, operation)
+    elif isinstance(item, dict):
+        aggregation[operation].append(item)
+    else:
+        aggregation[operation].append(_parse_term(item))
+
+
+def _parse_term(item):
+    """
+    parse term, on this level can be implemented rules according to range, term, match and others
+    https://www.elastic.co/guide/en/elasticsearch/reference/2.1/term-level-queries.html
+    :param item: string
+    :return: dict which contains {'term': {field_name: field_value}
+    """
+
+    field, value = smart_split(item)
+
+    if '|' in value:
+        values = value.split('|')
+        return {'bool': {'should': [{'term': {field: value}} for value in values]}}
+    if value == '_missing_':
+        return {'missing': {'field': field}}
+    if value.startswith('[') and value.endswith(']'):
+        return _parse_range(field, value[1:len(value) - 1])
+    return {'term': {field: value}}
+
+
+def _parse_range(field, value):
+    """
+    convert date range to ES range query.
+    https://www.elastic.co/guide/en/elasticsearch/reference/2.1/query-dsl-range-query.html
+    :param field: string, searched field name
+    :param value: string, date range, example [2016-07-10T00:00:00 TO 2016-08-10T01:00:00]
+    :return: dict, {'range': {field_name: {'gte': 2016-07-10T00:00:00, 'lte': 2016-08-10T01:00:00}}
+    """
+
+    from_, to = list(map(lambda string: string.strip(), value.split('TO')))
+    range_ = {'range': {field: {}}}
+
+    if from_ != '_missing_':
+        range_['range'][field].update({'gte': from_})
+
+    if to != '_missing_':
+        range_['range'][field].update({'lte': to})
+
+    return range_
+
+
+# attach _nested to nested_document
+def _parse_nested_items(query_string):
+    """
+    attach _nested to nested_document
+    :param query_string: string
+    :return: string with updated name for nested document, like assignments_nested for assignments
+    """
+    parsed_query_string = ''
+    in_quotes = False
+    for index, key in enumerate(query_string):
+
+        if key == '"' and not in_quotes:
+            in_quotes = True
+            key = ''
+        elif key == '"' and in_quotes:
+            in_quotes = False
+            key = ''
+
+        if key == '.' and not in_quotes:
+            key = '_nested.'
+
+        parsed_query_string = parsed_query_string + key
+    return parsed_query_string
+
+
+def _is_nested(item):
+    if isinstance(item, str):
+        field, _ = smart_split(item)
+        return '_nested' in field
+    return False
+
+
+def smart_split(item, split_key=':'):
+    """
+    split string in first matching with key
+    :param item: string which contain field_name:value or field_name:[00:00:00 TO 01:00:00]
+    :param split_key: key, which we use to split string
+    :return:
+    """
+    split_index = -1
+    for index, key in enumerate(item):
+        if key == split_key:
+            split_index = index
+            break
```

Comment:
This is essentially what str.find() does:
https://docs.python.org/3.4/library/stdtypes.html#str.find